### PR TITLE
Added basic color picker

### LINKED ISSUE #7996: Color picker improvements for easier text color styling
See [Figma](https://www.figma.com/file/1YM3jA9h4QGc4xLdyaXIJS/Stories-Sprint-1.7?node-id=9078%3A353191) for design.

## Description

We're reworking the color picker for making it easier for the users to choose colors.

- Color picker's default view will be with default preset colors with saved color below (view 1).
Screenshot 2021-07-02 at 15 21 48
- The Saved colors are only the global colors. Current story colors feature will remain but will be usable only from the panel.
- If the user wants to choose a custom color, the "Custom" button needs to be clicked.
- When clicking the Custom button, the current color picker view (view 2) will appear:
Screenshot 2021-07-02 at 15 24 03
- The user can choose a color from there and also use the "Add" button to save the color to _global_ colors.
- Note: If the user has selected an element that already has gradient assigned, and opens the color picker then, the View 2 will be opened by default right away.

### PR DESCRIPTION
**Pull request by barklund**
## Context

This adds a new basic color picker.

There are a few variants depending on the settings:

* For shape colors, everything is allowed
* For page background, opacity is not allowed
* For font and border colors, gradients are not allowed
* For text backgrounds, saved colors are not allowed

## To-do

* [x] Actually save the custom color
* [x] Maybe fine-tune the designs a bit (the saved color swatches don't align with the others)
* [x] What to display when there are no saved colors
* Better keyboard interaction ➡️    #8657 
* Integration tests ➡️    #8655 
* Saved color overflow ➡️    #8656 

### Page background flow – observe no (semi)transparency:
bcp-background

### Text color flow - observe no gradients:
bcp-text

### Text background flow - observe no saved colors nor opportunity to save:
bcp-text-bg

### Shape flow - observe that the shape already had a gradient, so color picker starts at custom view:
bcp-shape

### QA

This PR can be tested by following these steps:

#### Shape color
1. Add a new shape
2. Click to select the shape color
3. Observe the new basic color picker
4. Click to select a custom color
5. Observe the custom color picker working just as before
6. Click the back arrow in the top left corner
7. Observe that you're back at the basic color picker
8. Pick colors in both of the two pickers and observe the shape correctly updating

#### Page background
1. Add a new page
2. Click to select the page background-color
3. Observe the new basic color picker with the first color, transparent, being inactive
4. Click to select a custom color
5. Observe that the opacity input and track is not displayed

#### Font color
1. Add a text field
2. Click to select the font color
3. Observe the new basic color picker
4. Click to select a custom color
5. Observe that the gradient type picker is not displayed

#### Text background-color
1. Add a text field
2. Add text background fill
2. Click to select the text background-color
3. Observe the new basic color picker without saved colors at the bottom
4. Click to select a custom color
5. Observe that the add button is not displayed

### UAT

- [x] UAT should use the same steps as above.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #7996

### CODE REVIEW

**Comment by swissspidy** [comment]
Could you add some screenshots for the user-facing changes? Thanks :)

**Comment by miina** [comment]
One note from the requirements:
-  If the user has selected an element with gradient (non-solid color) and opens the color picker, the Custom color view will be displayed right away.

Looks like currently it's always opening the default view.

File: assets/src/edit-story/components/panels/design/shapeStyle/shapeStyle.js [hunk 1]
```diff
@@ -54,12 +53,11 @@ function ShapeStylePanel({ selectedElements, pushUpdate }) {
       <Row>
         <Color
           ref={colorInputRef}
-          hasGradient
+          allowsGradient
           value={backgroundColor}
           isMultiple={backgroundColor === ''}
           onChange={onChange}
           label={__('Background color', 'web-stories')}
-          hasOpacity={!isBackground}
```
Comment (miina) [inline assets/src/edit-story/components/panels/design/shapeStyle/shapeStyle.js]: Not seeing the opacity field anymore in the design panel for shapes, probably related to removal of this line?
Screenshot 2021-08-03 at 16 49 59

File: assets/src/edit-story/components/panels/design/preset/colorPreset/colorPresetActions.js [hunk 2]
```diff
@@ -17,241 +17,64 @@
...
-    }) => ({
-      selectedElements,
-      globalStoryStyles,
-      currentStoryStyles,
-      updateStory,
-    })
+    }) => ({ globalStoryStyles })
   );
 
   const { colors: globalColors } = globalStoryStyles;
```
Comment (miina) [inline assets/src/edit-story/components/panels/design/preset/colorPreset/colorPresetActions.js]: Use `colors: savedColors` directly?

File: assets/src/edit-story/components/colorPicker/patternTypePicker.js [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    [setToGradient]
+  );
+
+  return (
+    <Wrapper>
+      <ToggleButton
+        aria-label={__('Solid pattern type', 'web-stories')}
+        onClick={setToSolid}
+        size={BUTTON_SIZES.SMALL}
+        variant={BUTTON_VARIANTS.SQUARE}
```
Comment (miina) [inline assets/src/edit-story/components/colorPicker/patternTypePicker.js]: Hm, these are using some other style than usual per Figma:
Screenshot 2021-08-03 at 16 47 59

They have just a border around instead of background as usually for buttons.

**Review by miina — COMMENTED** [review]
(Couldn't fully review yet, will continue a bit later)

File: assets/src/edit-story/components/colorPicker/patternTypePicker.js (hunk 3 above)
Comment (barklund) [reply assets/src/edit-story/components/colorPicker/patternTypePicker.js]: I went with the design system buttons, as I've previously been told, that the design system trumps any individual designs. If a button/component/icon/color doesn't exist in the base system, it shouldn't be used (unless told otherwise).

Yes, I agree that it looks different, but it's the only button component we have.

File: assets/src/edit-story/components/panels/design/preset/colorPreset/colorPresetActions.js (hunk 2 above)
Comment (barklund) [reply assets/src/edit-story/components/panels/design/preset/colorPreset/colorPresetActions.js]: Yeah sure, I was just cleaning this up roughly removing all the unnecessary, not actually reviewing the code to see if it could be optimized. I'll give this an extra overhaul now that it's been greatly simplified.

File: assets/src/edit-story/components/panels/design/shapeStyle/shapeStyle.js (hunk 1 above)
Comment (barklund) [reply assets/src/edit-story/components/panels/design/shapeStyle/shapeStyle.js]: Oh! Well spotted - I wondered where that input went 😂

File: assets/src/edit-story/components/colorPicker/patternTypePicker.js (hunk 3 above)
Comment (BrittanyIRL) [reply assets/src/edit-story/components/colorPicker/patternTypePicker.js]: if we need to add a button variant I think we should. I've had this happen a few times now where the designs for the ds were just incomplete, which is fine, that should be a living system.

**Comment by barklund** [comment]
> * If the user has selected an element with gradient (non-solid color) and opens the color picker, the Custom color view will be displayed right away.

I was a bit in doubt about how to handle this.

- Should the custom picker view in case of a gradient then not have the back button in the top left corner?
- What if you then switch to a solid color, does the back button then come back?
- What happens if you apply a gradient from your saved colors in the basic color view - should it then switch to the custom color view instantly?

I didn't have a good place to ask these questions, so I think it makes sense to file as a follow-up when we have a UX resource for this kind of behavior. But I'll add it to the todo above!

**Comment by miina** [comment]
> - Should the custom picker view in case of a gradient then not have the back button in the top left corner?
> - What if you then switch to a solid color, does the back button then come back?

Everything else should stay the same, just the opened view should be the custom color view.

> - What happens if you apply a gradient from your saved colors in the basic color view - should it then switch to the custom color view instantly?

It should stay the same, this requirement was just about opening the color picker.

Edit: Let's skip if you think this behaviour might be confusing and needs additional UX confirmation.

File: assets/src/edit-story/components/colorPicker/patternTypePicker.js (hunk 3 above)
Comment (miina) [reply assets/src/edit-story/components/colorPicker/patternTypePicker.js]: This looks the most similar to what's in the design:
Screenshot 2021-08-04 at 09 36 45

It might need just making sure that the active button has the correct border color (in case it doesn't already).

**Comment by miina** [comment]
Looks like there's one additional change in Figma, for the Eyedropper button and the hex/alpha inputs -- they're aligned left now and the eyedropper seems to be a `quaternary` button, too:
Screenshot 2021-08-04 at 09 58 43

File: packages/story-editor/src/components/colorPicker/patternTypePicker.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import {
+  ToggleButton,
+  Icons,
+  BUTTON_SIZES,
+  BUTTON_VARIANTS,
+  BUTTON_TYPES,
+} from '@web-stories-wp/design-system';
+
+/**
+ * Internal dependencies
```
Comment (swissspidy) [inline packages/story-editor/src/components/colorPicker/patternTypePicker.js]: Can remove this comment

File: packages/story-editor/src/components/form/color/colorInput.js [hunk 5]
```diff
@@ -224,17 +224,15 @@ const ColorInput = forwardRef(function ColorInput(
         placement={PLACEMENT.LEFT_START}
         spacing={spacing}
         invisible={isEyedropperActive}
-        renderContents={({ propagateDimensionChange }) => (
```
Comment (miina) [inline packages/story-editor/src/components/form/color/colorInput.js:227]: Hm, I think the propagation might still be needed, for example when going from view 2 to the Default view, the lower part can get hidden otherwise:
propagation

File: packages/story-editor/src/components/colorPicker/test/colorPicker/moveStopByPointer.js [hunk 6]
```diff
@@ -49,17 +49,17 @@ describe('<ColorPicker /> when moving a stop with a pointer device', () => {
       clientX: 0,
     });
 
-    // Move to the 20% mark
+    // Move to the 30% mark
```
Comment (miina) [inline packages/story-editor/src/components/colorPicker/test/colorPicker/moveStopByPointer.js:52]: Curious -- why the change?

File: packages/story-editor/src/components/form/color/colorInput.js (hunk 5 above)
Comment (miina) [reply packages/story-editor/src/components/form/color/colorInput.js:227]: Q: Previously, the Saved colors had a scrollbar if more then 3 lines, should it be the same now, too?
Screenshot 2021-08-12 at 12 28 27

**Review by miina — APPROVED** [review]
The buttons look great now!

Noticed just one issue with the color picker dimensions.

**Comment by swissspidy** [comment]
> Could you add some screenshots for the user-facing changes? Thanks :)

Not to be pedantic, but there's a `user-facing changes` section in the PR template for this reason :) Would be useful to have an overview of the changes to this component at first glance

File: packages/story-editor/src/components/colorPicker/test/colorPicker/moveStopByPointer.js (hunk 6 above)
Comment (barklund) [reply packages/story-editor/src/components/colorPicker/test/colorPicker/moveStopByPointer.js:52]: Rounding differences due to the decreased width of the entire bar - the previous 20% rounded to 21% because 🪄 .

File: packages/story-editor/src/components/form/color/colorInput.js (hunk 5 above)
Comment (barklund) [reply packages/story-editor/src/components/form/color/colorInput.js:227]: > Hm, I think the propagation might still be needed, for example when going from view 2 to the Default view, the lower part can get hidden otherwise:

Dammit, I hoped you wouldn't notice! 😄 

I did see this one time but forgot again. I will file as a follow-up bug, as it's pretty low priority.

> Q: Previously, the Saved colors had a scrollbar if more then 3 lines, should it be the same now, too?

Will be fixed in #8656 (only two lines to be shown).

**Comment by barklund** [comment]
> > Could you add some screenshots for the user-facing changes? Thanks :)
> 
> Not to be pedantic, but there's a `user-facing changes` section in the PR template for this reason :) Would be useful to have an overview of the changes to this component at first glance

Oh, right. Will add some gifs for the various configs of this!

File: packages/story-editor/src/components/form/color/colorInput.js (hunk 5 above)
Comment (barklund) [reply packages/story-editor/src/components/form/color/colorInput.js:227]: > > Hm, I think the propagation might still be needed, for example when going from view 2 to the Default view, the lower part can get hidden otherwise:
> 
> I will file as a follow-up bug, as it's pretty low priority.

Filed in #8673

Comment (merapi) [reply packages/story-editor/src/components/form/color/colorInput.js:227]: Looks like it will be handled in #8656 (via 3 dots button, max 2 rows)
image
